askametric/query_processor/query_processor.py

```diff
@@ -63,128 +91,279 @@ async def _get_query_language_from_llm(self) -> None:
         """
-        sql_result = await self.tools.run_sql(self.sql_query, self.asession)
+        sql_result = await self.tools.run_sql(
+            self.user_query_response.sql_query, self.asession
+        )
         prompt = create_final_answer_prompt(
-            self.query,
-            self.sql_query,
+            self.updated_query,
+            self.user_query_response.sql_query,
             sql_result,
-            self.query_language,
-            self.query_script,
+            self.user_query_response.query_language,
+            self.user_query_response.query_script,
         )
         final_answer_llm_response = await _ask_llm_json(
-            prompt, self.system_message, llm=self.llm, temperature=self.temperature
+            prompt=prompt,
+            system_message=self.system_message,
+            context_message=self.user_query_response.conversation_summary,
+            llm=self.llm,
+            temperature=self.temperature,
         )
 
-        self.final_answer = final_answer_llm_response["answer"]["answer"]
+        self.user_query_response.text_response = final_answer_llm_response["answer"][
+            "answer"
+        ]
         self.cost += float(final_answer_llm_response["cost"])
-        self.final_answer_prompt = prompt
 
     @track_time(create_class_attr="timings")
-    async def process_query(self) -> None:
+    async def process_query(self, query: dict) -> None:
         """
         The function processes the user query and returns the final answer.
+
+        Args:
+            query: The user's query. It is a dictionary with the
+                following keys: query_text, query_metadata.
         """
+        # --- Create / overwrite class attributes ---
+        self.query = query
+        self.cost = 0.0
+        self.user_query_response = UserQueryResponse()
+        self.guardrails = LLMGuardRails(self.llm, self.system_message)
+        self.prompts = Prompts()
+
+        # --- Begin processing the query ---
         # Get query language
         await self._get_query_language_from_llm()
 
-        # Check query safety
+        # Translate query into English to be used for LLM's processing steps
+        self.eng_translation = await self._english_translation()
+
+        # Check query for code
+        await self.guardrails.check_code(query=self.eng_translation["query_text"])
+        if self.guardrails.code is True:
+            self.user_query_response.text_response = self.guardrails.code_response
+            self._update_results()
+
+            return None
+
+        # Get context for previous queries
+        await self._get_conversation_summary_from_llm()
+        self.updated_query = {
+            "query_text": self.user_query_response.updated_query_text,
+            "query_metadata": self.eng_translation["query_metadata"],
+        }
+
+        if self.user_query_response.text_response != "":
+            self._update_results()
+
+            return None
```

Comment:
What do you think about running all the safety checks/rails even if we were able to generate response from past data? It will just be safer. Otherwise, we could have potential safety loopholes. Eg, You could answer an unsafe question via a combo of past questions. Don't know what would look like tho. But easier to be safer.

Reply:
For now, moved the return statement (if the final answer is not empty) to after the guardrail checks. At least we won't return an answer if it doesn't pass the checks. Will